### PR TITLE
[Stdlib] Eliminate the _UnicodeEncoding_ protocol.

### PR DESCRIPTION
**Pull request by DougGregor**
This protocol was only used to fake recursive protocol conformances;
collapse it into `_UnicodeEncoding`.

### CODE REVIEW

**Comment by DougGregor** [comment]
This builds more on top of https://github.com/apple/swift/pull/11751.

**Comment by DougGregor** [comment]
@swift-ci please test Linux

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: lib/AST/GenericSignatureBuilder.cpp [hunk 1]
```diff
@@ -4963,18 +5146,380 @@ namespace {
...
+/// Collapse same-type components within an equivalence class, minimizing the
+/// number of requirements required to express the equivalence class.
+static void collapseSameTypeComponents(
+              GenericSignatureBuilder &builder,
+              EquivalenceClass *equivClass,
+              llvm::SmallDenseMap<PotentialArchetype *, unsigned> &componentOf,
+              std::vector<IntercomponentEdge> &sameTypeEdges) {
+  SmallVector<unsigned, 4> collapsedParents;
+  for (unsigned i : indices(equivClass->derivedSameTypeComponents))
+    collapsedParents.push_back(i);
```
Comment (moiseev) [inline lib/AST/GenericSignatureBuilder.cpp]: Aren't implicit blocks considered evil?

Comment (DougGregor) [reply lib/AST/GenericSignatureBuilder.cpp]: I, personally, don't consider them evil... am I committing a sin?

Comment (moiseev) [reply lib/AST/GenericSignatureBuilder.cpp]: I'm just thinking about the "gotofail".

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: lib/AST/GenericSignatureBuilder.cpp (hunk 1 above)
Comment (jrose-apple) [reply lib/AST/GenericSignatureBuilder.cpp]: I swear it used to be in the LLVM conventions, but now I don't see it and didn't see an obvious commit that removed it.

Comment (DougGregor) [reply lib/AST/GenericSignatureBuilder.cpp]: I'll just fix it

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
Yippee! I'm going to fold this into https://github.com/apple/swift/pull/11852, because I have too many interrelated PRs going on and this isolated standard library change isn't a user-visible thing anyway.

**Pull request closed without merging**